ts/Series/SeriesOnPointComposition.ts

```diff
@@ -0,0 +1,576 @@
+     *
+     * @ignore
+    */
+    function afterInit(this: Series): void {
+        if (this.options.onPoint) {
+            this.onPoint = new Additions(this as SeriesComposition);
+        }
+    }
+
+    /**
+     * Fire series.getRadii method to calculate required z data before original
+     * translate.
+     *
+     * @ignore
+     * @function Highcharts.Series#translate
+    */
+    function seriesTranslate(this: Series): void {
+        this.onPoint && this.onPoint.getRadii();
+
+        translateFunctions[this.type + 'Translate'].call(this);
+    }
+
+    /**
+     * Recalculate series.center (x, y and size).
+     *
+     * @ignore
+     * @function Highcharts.Series#getCenter
+    */
+    function seriesGetCenter(this: Series): Array<number> {
+        const onPointOptions = this.options.onPoint;
+
+        let ret = getCenter.call(this);
+
+        if (onPointOptions) {
+            const connectedPoint = this.chart.get(onPointOptions.id);
+
+            if (
+                connectedPoint instanceof Point &&
+                defined(connectedPoint.plotX) &&
+                defined(connectedPoint.plotY)
+            ) {
+                ret[0] = connectedPoint.plotX;
+                ret[1] = connectedPoint.plotY;
+            }
+
+            const position = onPointOptions.position;
+
+            if (position) {
+                if (defined(position.x)) {
+                    ret[0] = position.x;
+                }
+
+                if (defined(position.y)) {
+                    ret[1] = position.y;
+                }
+
+                if (position.offsetX) {
+                    ret[0] += position.offsetX;
+                }
+
+                if (position.offsetY) {
+                    ret[1] += position.offsetY;
+                }
+            }
+        }
+
+        // Get and set the size
+        const radius = this.radii && this.radii[this.index];
+
+        if (isNumber(radius)) {
+            ret[2] = radius * 2;
+        }
+
+        return ret;
+    }
+
+    function seriesDrawPoints(this: Series): void {
+        drawPointsFunctions[this.type + 'DrawPoints'].call(this);
+
+        this.onPoint && this.onPoint.seriesDrawConnector();
+    }
+
+    /* *
+     *
+     *  Classes
+     *
+     * */
+
+    /**
+     * @private
+     */
+    export class Additions {
+
+        /* *
+         *
+         *  Constructors
+         *
+         * */
+
+        /**
+         * @private
+         */
+        public constructor(series: SeriesComposition) {
+            this.series = series;
+            this.options = series.options.onPoint;
+        }
+
+        /* *
+         *
+         *  Properties
+         *
+         * */
+
+        public series: SeriesComposition;
+
+        public connector?: SVGElement;
+
+        public options?: OnPoint;
+
+        /**
+         * @ignore
+         */
+        public getRadii(): void {
+            bubble.prototype.getRadii.call(this.series);
```

Comment:
If we copied over `zData`, `yData` and other properties required to calculate the radii to the composition instance - cloud we then avoid calling this with the main series scope? So we could simply do this:

`public getRadii = bubble.prototype.getRadii;`

And the same for `getRadius`, `getPxExtremes` and `getZExtremes`.

Reply:
I did my best, but I couldn't avoid the context swap in `BubbleSeries.ts` in some cases.

Also, I still needed to save `radii` on the `series` (`series.onPoint.radii` isn't enough)